elastic/kibana
# [Lens] Time shift metrics

Fixes https://github.com/elastic/kibana/issues/68458

Adds the ability to shift individual metrics in time. The shift can either be a unit of time (`1d`, `6h`, ...) or `previous`, which will shift the metric by one date bucket if there's a date histogram or by the length of the current time range if there is no date histogram in the current chart.

As long as all metrics in a chart are shifted by the same amount of time, there is no limitation in configuring it - however if there are different shifts (like no shift and a shift by one day), the terms function is not allowed. In this case an error message will the shown and the user can resolve it by pinning the current top values, creating a filters function:

Single shifted metric with terms - works fine:
<img width="1044" alt="Screenshot 2021-05-05 at 09 38 24" src="https://user-images.githubusercontent.com/1508364/117110390-55b3c600-ad86-11eb-8951-af60cf9a69d3.png">

Adding another unshifted metric leads to an error:
<img width="1195" alt="Screenshot 2021-05-05 at 09 38 36" src="https://user-images.githubusercontent.com/1508364/117110434-649a7880-ad86-11eb-90a8-364b89cd8852.png">

Can be "fixed" by pinning the current top values:
<img width="1167" alt="Screenshot 2021-05-05 at 09 38 52" src="https://user-images.githubusercontent.com/1508364/117110490-78de7580-ad86-11eb-9a8a-c304ac9f4c48.png">

For the general approach check out the request/response examples in https://gist.github.com/flash1293/d9acacdd9c6439c1cf507f92b05ccca9

I’m going to add more cases to the functional testing, but the code can be reviewed already.

## Code changes AggConfigs

* All metrics get a timeshift param
* Move logic to construct time filter for request into agg configs (`getSearchSourceTimeFilter`). If there are shifted metrics, build an OR query for all time ranges
* Add post flight transform to agg configs which is called by search source after the response is returned from ES. This is used to merge values from different time shifts (will recursively step through all bucket layers until it reaches the filters split `transformTimeShift` - then will merge together the splitted branches into a single one recursively `mergeAggLevel`) **This is the core part of the logic**
* Extend agg type by `splitForTimeShift` function which will return true if a filters split should be inserted above the dsl for this agg config
* Extend bucket agg type with new methods: `getShiftedKey` (to shift a bucket key, used in date histogram), `orderBuckets` (to order merged buckets if some only existed in one branch of the time split), `getTimeShiftInterval` (to get the underlying interval of the request for powering the "previous" time shift mode, used in date histogram)
* Add logic to `toDsl` include time filter split in the request at the right place

## Code changes Lens

* Add a UI to configure time shift similar to filter UI
* Add additional validation to catch error cases
* Extend error messages with an optional fix action to provide the "Pin top values" functionality
* Add timeShift parameter to the expression ast of the datasource

## Testing the agg config level (without Lens)

To test the agg config level:
* Start the functional tests server for interpreter tests (this is also what's used for automated testing): `node scripts/functional_tests_server.js --config test/interpreter_functional/config.ts`
* Go to `http://localhost:5620`
* Install some sample data
* Go to Stack management and look up id of index pattern
* Go to "Run Pipeline" app
* Open console
* Run expressions like this:
```
await window.runPipeline(`kibana_context timeRange={timerange from='2021-05-07T00:00:00.000Z' to='2021-05-11T00:00:00.000Z'}
                      | esaggs index={indexPatternLoad id='90943e30-9a47-11e8-b64d-95841ca0b247'}
                              aggs={aggTerms id="1" field="geo.src" size="3" enabled=true schema="bucket" orderAgg={aggCount id="order" enabled=true schema="metric"} otherBucket=true}
          aggs={aggAvg id="2" field="bytes" enabled=true schema="metric" timeShift="1d"}
          aggs={aggAvg id="3" field="bytes" enabled=true schema="metric"}`, { type: 'null' })
```
* Check whether the result is correct

## Testing on Lens level

For each metric dimension, theres a "Shift in time" option under "Advanced options"
